### ISSUE TITLE
Caffe Converter computes wrong results. Simply not working.

### ISSUE THREAD

**Issue by rmbarbosa** [tiny-dnn/tiny-dnn#429]
Hello,
I compiled successfully the protobuf, and tiny-cnn and caffe_converter using VStudio 2015 64 bit versions.
But i have a problema with the results of the network.
Apparently it loads well the network and converts it successfully but the results are completly diferent.

using tiny-cnn I run the command:
**_caffe_converter.exe deploy.prototxt bvlc_alexnet.caffemodel imagenet_mean.binaryproto synset_words.txt cat.jpg_**
and the results are:
Elapsed time(s): 2.113
n02090721 Irish wolfhound,0.934235
n03814906 necklace,0.0107539
n03976657 pole,0.00875462
n03124043 cowboy boot,0.00467396
n03063689 coffeepot,0.00278851

but when i run the caffe version using the same data, i get:

**_classification.exe deploy.prototxt bvlc_alexnet.caffemodel imagenet_mean.binaryproto synset_words.txt cat.jpg_**
---------- Prediction for cat.jpg ----------
0.3094 - "n02124075 Egyptian cat"
0.1761 - "n02123159 tiger cat"
0.1221 - "n02123045 tabby, tabby cat"
0.1132 - "n02119022 red fox, Vulpes vulpes"
0.0421 - "n02085620 Chihuahua"

I've tried to see if the problema was related with the imagetype but its defined as image_type::bgr and it should be ok.
i've also tried with some samples from the mnist exemple and changed the image_type to grayscale but the problem is the same: the results are wrong when using the tiny-cnn converter.

Can anyone help? i'm i doing something wrong? did anybody got the same results?

thanks in advance.

**Comment by zjulujia** [comment]
I've got similar results as you, do you figure it out?

`./example_caffe_converter ~/works/caffe/models/bvlc_reference_caffenet/deploy.prototxt ~/works/caffe/models/bvlc_reference_caffenet/bvlc_reference_caffenet.caffemodel ~/works/caffe/data/ilsvrc12/imagenet_mean.binaryproto ~/works/caffe/data/ilsvrc12/synset_words.txt ~/works/caffe/examples/images/cat.jpg
[INFO] convert Convolution => N8tiny_dnn19convolutional_layerINS_10activation8identityEEE
[INFO] shape:55x55x96
[INFO] convert ReLU => N8tiny_dnn12linear_layerINS_10activation4reluEEE
[INFO] shape:55x55x96
[INFO] convert Pooling => N8tiny_dnn17max_pooling_layerINS_10activation8identityEEE
[INFO] shape:27x27x96
[INFO] convert LRN => N8tiny_dnn9lrn_layerINS_10activation8identityEEE
[INFO] shape:27x27x96
[INFO] convert Convolution => N8tiny_dnn19convolutional_layerINS_10activation8identityEEE
[INFO] shape:27x27x256
[INFO] convert ReLU => N8tiny_dnn12linear_layerINS_10activation4reluEEE
[INFO] shape:27x27x256
[INFO] convert Pooling => N8tiny_dnn17max_pooling_layerINS_10activation8identityEEE
[INFO] shape:13x13x256
[INFO] convert LRN => N8tiny_dnn9lrn_layerINS_10activation8identityEEE
[INFO] shape:13x13x256
[INFO] convert Convolution => N8tiny_dnn19convolutional_layerINS_10activation8identityEEE
[INFO] shape:13x13x384
[INFO] convert ReLU => N8tiny_dnn12linear_layerINS_10activation4reluEEE
[INFO] shape:13x13x384
[INFO] convert Convolution => N8tiny_dnn19convolutional_layerINS_10activation8identityEEE
[INFO] shape:13x13x384
[INFO] convert ReLU => N8tiny_dnn12linear_layerINS_10activation4reluEEE
[INFO] shape:13x13x384
[INFO] convert Convolution => N8tiny_dnn19convolutional_layerINS_10activation8identityEEE
[INFO] shape:13x13x256
[INFO] convert ReLU => N8tiny_dnn12linear_layerINS_10activation4reluEEE
[INFO] shape:13x13x256
[INFO] convert Pooling => N8tiny_dnn17max_pooling_layerINS_10activation8identityEEE
[INFO] shape:6x6x256
[INFO] convert InnerProduct => N8tiny_dnn21fully_connected_layerINS_10activation8identityEEE
[INFO] shape:4096x1x1
[INFO] convert ReLU => N8tiny_dnn12linear_layerINS_10activation4reluEEE
[INFO] shape:4096x1x1
[INFO] convert Dropout => N8tiny_dnn13dropout_layerE
[INFO] shape:4096x1x1
[INFO] convert InnerProduct => N8tiny_dnn21fully_connected_layerINS_10activation8identityEEE
[INFO] shape:4096x1x1
[INFO] convert ReLU => N8tiny_dnn12linear_layerINS_10activation4reluEEE
[INFO] shape:4096x1x1
[INFO] convert Dropout => N8tiny_dnn13dropout_layerE
[INFO] shape:4096x1x1
[INFO] convert InnerProduct => N8tiny_dnn21fully_connected_layerINS_10activation8identityEEE
[INFO] shape:1000x1x1
[INFO] convert Softmax => N8tiny_dnn12linear_layerINS_10activation7softmaxEEE
[INFO] shape:1000x1x1
Elapsed time(s): 0.891706
n02090721 Irish wolfhound,0.870049
n03976657 pole,0.0188536
n03814906 necklace,0.0151203
n03124043 cowboy boot,0.00946595
n03063689 coffeepot,0.00499552
`

**Comment by edgarriba** [comment]
@rmbarbosa could you check if with https://github.com/tiny-dnn/tiny-dnn/tree/v0.1.1 you still get the same results?

**Comment by rmbarbosa** [comment]
Hello @edgarriba.
I've managed to compile it using v0.1.1 and **it works**. the results are exactly the ones from caffe.

But this is definitely a problem with v1.0.0xx because I've compiled the caffe_converter.cpp from v0.1.1 but using the v1.0.0xx with OpenCV dependency (considering that it could be something to do with the image processing) and the results are also wrong. So caffe_converter compiled against v1.0.0.xx fails, and against v0.1.1 works.

for the record, the result of caffe converter compiled against v0.1.1 are exactly the same as the caffe output.
**_caffe_converter.exe deploy.prototxt bvlc_alexnet.caffemodel imagenet_mean.binaryproto synset_words.txt cat.jpg_**
Elapsed time(s): 1.907
n02124075 Egyptian cat,0.309396
n02123159 tiger cat,0.176119
n02123045 tabby, tabby cat,0.122125
n02119022 red fox, Vulpes vulpes,0.113246
n02085620 Chihuahua,0.0421218

**Comment by edgarriba** [comment]
@rmbarbosa thx for the effort. Between v0.1.1 and v1.0.0 @nyanp introduced some changes to image API. Probably he's the correct person to find the issue

**Comment by shenyuhit** [comment]
@edgarriba I have used  v0.1.1  for predict , i use squeeze，but it does not work! Message is
 "terminate called after throwing an instance of 'std::runtime_error'
  what():  error: tiny-cnn does not support this layer type:Split"
. Did you have any advice!

**Comment by goranrauker** [comment]
The caffe converter first breaks in this commit
https://github.com/tiny-dnn/tiny-dnn/commit/5c2ab5b99abe4c283b290d9338aea080430261d7#diff-1ccb101c13602ae037f1d65ed2811dbd

**Comment by animetrics** [comment]
Is this just the caffe converter that's broken.  So if you were to convert a model via v0.1.1 converter, but then deploy it using v1.0.0 does it work properly?

**Comment by rperrot** [comment]
There's an index difference introduced here : 

https://github.com/tiny-dnn/tiny-dnn/commit/5c2ab5b99abe4c283b290d9338aea080430261d7#diff-1201f4bcc000cb58d71ee51947e42774R149

But I don't think it's the only thing that cause the issue.

**Comment by edgarriba** [comment]
hey guys thanks for your effort to fix this issue!
Maybe you want to join us in [gitter/developers](https://gitter.im/tiny-dnn/developers) room and share thoughts.

**Comment by animetrics** [comment]
I wish I had more initial experience with the library to help, but right now I'm just trying to vet whether or not it is suitable for my purposes.  It does seem like this is a significant issue as I imagine the library is quite useful for deploying networks that have been trained via caffe, or fine tuning such networks.  

I did try to use the converter from 0.1.1, save the model, and then import it in v1.0.0.  But it looks like the 0.1.1 serialization is primitive enough that to deserialize, you still need to separately do a manual network construction before reading in the weights.

**Comment by logic1988** [comment]
I encountered the same problem that the result of MNIST and caffe model were all incorrect in v1.0.0, while all right in v0.1.1.

**Comment by edgarriba** [comment]
@logic1988 notice that @goranrauker found that in commit https://github.com/tiny-dnn/tiny-dnn/issues/429#issuecomment-265820137 something is broken. If someone finds the issue don't hesitate to pull request it.

**Comment by edgarriba** [comment]
Guys check master again. It's supposed that the caffe converter is fixed now
https://github.com/tiny-dnn/tiny-dnn/pull/515

**Comment by animetrics** [comment]
That's great, but just pulled latest and it's better but not quite correct.  Scores are a bit off and 5th result should be lynx, catamount not kit fox.   I get,

    n02123045 tabby, tabby cat,0.331109
    n02123159 tiger cat,0.201654
    n02124075 Egyptian cat,0.148082
    n02119022 red fox, Vulpes vulpes,0.122779
    n02119789 kit fox, Vulpes macrotis,0.0714079

but expecting

    n02123045 tabby, tabby cat,0.313394
    n02123159 tiger cat,0.237981
    n02124075 Egyptian cat,0.1235
    n02119022 red fox, Vulpes vulpes,0.100342
    n02127052 lynx, catamount,0.0715481

**Comment by edgarriba** [comment]
where the expected results come from?
https://github.com/tiny-dnn/tiny-dnn/commit/5c2ab5b99abe4c283b290d9338aea080430261d7#diff-1201f4bcc000cb58d71ee51947e42774R149

**Comment by animetrics** [comment]
Running v0.1.1 for [this example](https://github.com/BVLC/caffe/tree/master/examples/cpp_classification), which is the one referenced in the caffe_converter readme, and which matches exactly the results there.

**Comment by edgarriba** [comment]
ok. A bit difficult to find the issue since it's a precision issue. /cc @beru any idea?

**Comment by beru** [comment]
@edgarriba I wonder if switching to internal backend will resolve the issue.

**Comment by beru** [comment]
The backend name `Internal` is weird. Shouldn't it be `Reference`?
